File: altgraph_data_app/src/main/java/org/cjoakim/cosmos/altgraph/data/common/graph/v2/JGraphBuilder.java
package org.cjoakim.cosmos.altgraph.data.common.graph.v2;

import com.azure.cosmos.*;
import com.azure.cosmos.models.CosmosQueryRequestOptions;
import com.azure.cosmos.models.FeedResponse;
import lombok.extern.slf4j.Slf4j;
import org.cjoakim.cosmos.altgraph.data.DataAppConfiguration;
import org.cjoakim.cosmos.altgraph.data.common.DataConstants;
import org.cjoakim.cosmos.altgraph.data.common.io.JsonLoader;
import org.cjoakim.cosmos.altgraph.data.common.model.imdb.IndexDocument;
import org.cjoakim.cosmos.altgraph.data.common.model.imdb.Movie;
import org.cjoakim.cosmos.altgraph.data.common.util.MemoryStats;
import org.jgrapht.graph.DefaultEdge;
import org.jgrapht.graph.DirectedMultigraph;
import org.jgrapht.graph.Multigraph;

import java.util.HashMap;
import java.util.Iterator;

/**
 * This class builds/refreshes an in-memory Google Guava Graph "network" (i.e. - graph)
 * from either a disk or CosmosDB datasource.
 * <p>
 * It is invoked from the refresh() method of class JGraph.
 * <p>
 * Chris Joakim, Microsoft, November 2022
 */

@Slf4j
public class JGraphBuilder implements DataConstants {

    private String uri;
    private String key;
    private String dbName;
    private String source = null;
    private boolean directed = false;

    public JGraphBuilder(String source) {

        super();
        this.source = source;
    }

    public org.jgrapht.Graph<String, DefaultEdge> buildImdbGraph() throws Exception {

        log.warn("buildImdbGraph, source:   " + source);
        log.warn("buildImdbGraph, directed: " + directed);

        try {
            switch (source) {
                case IMDB_GRAPH_SOURCE_DISK:
                    return loadImdbGraphFromDisk(directed);
                case IMDB_GRAPH_SOURCE_COSMOS:
                    return loadImdbGraphFromCosmos(directed);
                default:
                    log.warn("undefined graph source: " + source);
            }
        } catch (Exception ex) {
            ex.printStackTrace();
            throw ex;
        }
        return null;
    }

    /**
     * See https://jgrapht.org/guide/UserOverview#graph-structures
     */
    private org.jgrapht.Graph<String, DefaultEdge> createGraphObject(boolean directed) {

        org.jgrapht.Graph<String, DefaultEdge> graph = null;

        if (directed) {
            graph = new DirectedMultigraph<>(DefaultEdge.class);
        } else {
            graph = new Multigraph<>(DefaultEdge.class);
        }
        return graph;
    }

    private org.jgrapht.Graph<String, DefaultEdge> loadImdbGraphFromDisk(boolean directed)
            throws Exception {

        org.jgrapht.Graph<String, DefaultEdge> graph = createGraphObject(directed);

        JsonLoader jsonLoader = new JsonLoader();
        HashMap<String, Movie> moviesHash = new HashMap<String, Movie>();
        jsonLoader.readMovieDocuments(moviesHash, true);
        checkMemory(true, true, "loadImdbGraphFromDisk - after reading movies from file");

        Iterator<String> moviesIt = moviesHash.keySet().iterator();
        long movieNodesCreated = 0;
        long personNodesCreated = 0;
        long edgesCreated = 0;

        while (moviesIt.hasNext()) {
            String tconst = moviesIt.next();
            Movie movie = moviesHash.get(tconst);
            if (!graph.containsVertex(tconst)) {
                graph.addVertex(tconst);
                movieNodesCreated++;
            }

            Iterator<String> peopleIt = movie.getPeople().iterator();
            while (peopleIt.hasNext()) {
                String nconst = peopleIt.next();
                if (!graph.containsVertex(nconst)) {
                    graph.addVertex(nconst);
                    personNodesCreated++;
                }
                graph.addEdge(nconst, tconst);  // person-to-movie
                edgesCreated++;
                if (directed) {
                    // just a single edge between vertices
                } else {
                    graph.addEdge(tconst, nconst);  // movie-to-person
                    edgesCreated++;
                }
            }
        }

        checkMemory(true, true, "loadImdbGraphFromDisk - after building graph");
        log.warn("loadImdbGraphFromDisk - movieNodesCreated:  " + movieNodesCreated);
        log.warn("loadImdbGraphFromDisk - personNodesCreated: " + personNodesCreated);
        log.warn("loadImdbGraphFromDisk - edgesCreated:       " + edgesCreated);
        return graph;
    }

    private org.jgrapht.Graph<String, DefaultEdge> loadImdbGraphFromCosmos(boolean directed) {

        uri = DataAppConfiguration.getInstance().uri;
        key = DataAppConfiguration.getInstance().key;
        dbName = DataAppConfiguration.getInstance().dbName;
        source = DataAppConfiguration.getInstance().imdbGraphSource;
        directed = DataAppConfiguration.getInstance().imdbGraphDirected;

        org.jgrapht.Graph<String, DefaultEdge> graph = createGraphObject(directed);

        CosmosAsyncClient client;
        CosmosAsyncDatabase database;
        CosmosAsyncContainer container;
        double requestCharge = 0;
        long documentsRead = 0;
        long movieNodesCreated = 0;
        long personNodesCreated = 0;
        long edgesCreated = 0;
        String sql = "select * from c where c.pk = 'movie_idx'";
        int pageSize = 1000;
        String continuationToken = null;
        CosmosQueryRequestOptions queryOptions = new CosmosQueryRequestOptions();

        log.warn("uri:    " + uri);
        log.warn("key:    " + key);
        log.warn("dbName: " + dbName);

        checkMemory(true, true, "loadImdbGraphFromCosmos - start");
        long startMs = System.currentTimeMillis();

        client = new CosmosClientBuilder()
                .endpoint(uri)
                .key(key)
                .preferredRegions(DataAppConfiguration.getPreferredRegions())
                .consistencyLevel(ConsistencyLevel.SESSION)
                .contentResponseOnWriteEnabled(true)
                .buildAsyncClient();

        database = client.getDatabase(this.dbName);
        log.warn("client connected to database Id: " + database.getId());

        container = database.getContainer(IMDB_INDEX_CONTAINER);
        log.warn("container: " + container.getId());

        long dbConnectMs = System.currentTimeMillis();

        try {
            do {
                Iterable<FeedResponse<IndexDocument>> feedResponseIterator =
                        container.queryItems(sql, queryOptions, IndexDocument.class).byPage(
                                continuationToken, pageSize).toIterable();  // Convert Asynch Flux to Iterable

                for (FeedResponse<IndexDocument> page : feedResponseIterator) {
                    for (IndexDocument doc : page.getResults()) {
                        documentsRead++;
                        if ((documentsRead % 10000) == 0) {
                            log.warn("" + documentsRead + " -> " + doc.asJson(false));
                        }
                        String tconst = doc.getTargetId();
                        graph.addVertex(tconst);
                        movieNodesCreated++;

                        for (int i = 0; i < doc.getAdjacentVertices().size(); i++) {
                            String nconst = doc.getAdjacentVertices().get(i);
                            if (!graph.containsVertex(nconst)) {
                                graph.addVertex(nconst);
                                personNodesCreated++;
                            }
                            graph.addEdge(nconst, tconst);  // person-to-movie
                            edgesCreated++;

                            if (directed) {
                                // just a single edge between vertices
                            } else {
                                graph.addEdge(tconst, nconst);  // movie-to-person
                                edgesCreated++;
                            }
                        }

                    }
                    requestCharge = requestCharge + page.getRequestCharge();
                    continuationToken = page.getContinuationToken();
                }
            }
            while (continuationToken != null);
        } catch (Throwable t) {
            t.printStackTrace();
        }
        long finishMs = System.currentTimeMillis();
        long dbConnectElapsed = dbConnectMs - startMs;
        long dbReadingElapsed = finishMs - dbConnectMs;
        double dbReadingSeconds = (double) dbReadingElapsed / 1000.0;
        long totalElapsed = finishMs - startMs;

        double ruPerSec = (double) requestCharge / dbReadingSeconds;

        checkMemory(true, true, "loadImdbGraphFromCosmos - after building graph");
        log.warn("loadImdbGraphFromCosmos - documentsRead:      " + documentsRead);
        log.warn("loadImdbGraphFromCosmos - movieNodesCreated:  " + movieNodesCreated);
        log.warn("loadImdbGraphFromCosmos - personNodesCreated: " + personNodesCreated);
        log.warn("loadImdbGraphFromCosmos - edgesCreated:       " + edgesCreated);
        log.warn("loadImdbGraphFromCosmos - requestCharge:      " + requestCharge);
        log.warn("loadImdbGraphFromCosmos - ru per second:      " + ruPerSec);
        log.warn("loadImdbGraphFromCosmos - db connect ms:      " + dbConnectElapsed);
        log.warn("loadImdbGraphFromCosmos - db read ms:         " + dbReadingElapsed);
        log.warn("loadImdbGraphFromCosmos - total elapsed ms:   " + totalElapsed);
        return graph;
    }

    protected MemoryStats checkMemory(boolean doGc, boolean display, String note) {
        if (doGc) {
            System.gc();
        }
        MemoryStats ms = new MemoryStats(note);
        if (display) {
            try {
                sysout(ms.asDelimitedHeaderLine("|"));
                sysout(ms.asDelimitedDataLine("|"));
            } catch (Exception e) {
                sysout("error serializing MemoryStats to JSON");
            }
        }
        return ms;
    }

    protected void sysout(String s) {
        System.out.println(s);
    }

}

File: src/altgraph_shared_app/Services/Graph/v2/JGraphBuilder.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using altgraph_shared_app.Models.Imdb;
using altgraph_shared_app.Options;
using Microsoft.Azure.Cosmos;
using Microsoft.Azure.Cosmos.Fluent;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using QuikGraph;

namespace altgraph_shared_app.Services.Graph.v2
{
  public class JGraphBuilder : IJGraphBuilder
  {
    public string Uri { get; set; } = string.Empty;
    public string Key { get; set; } = string.Empty;
    public string DbName { get; set; } = string.Empty;
    public string Source { get; set; } = string.Empty;
    public bool Directed { get; set; } = false;
    private ILogger<JGraphBuilder> _logger;
    private CosmosOptions _cosmosOptions;
    private ImdbOptions _imdbOptions;
    public event EventHandler<JGraphBuilderStartedLoadingProgressEventArgs>? JGraphBuilderStartedLoadingProgress;
    public event EventHandler<JGraphBuilderLoadingProgressEventArgs>? JGraphBuilderLoadingProgress;
    public event EventHandler<JGraphBuilderFinishedLoadingProgressEventArgs>? JGraphBuilderFinishedLoadingProgress;

    public JGraphBuilder(ILogger<JGraphBuilder> logger, IOptions<CosmosOptions> cosmosOptions, IOptions<ImdbOptions> imdbOptions)
    {
      _logger = logger;
      _cosmosOptions = cosmosOptions.Value;
      _imdbOptions = imdbOptions.Value;
      Source = _imdbOptions.GraphSource;
    }

    public async Task<IMutableGraph<string, Edge<string>>?> BuildImdbGraphAsync()
    {
      _logger.LogWarning($"buildImdbGraph, source:   {Source}");
      _logger.LogWarning($"buildImdbGraph, directed: {Directed}");

      try
      {
        switch (Source)
        {
          case Constants.IMDB_GRAPH_SOURCE_DISK:
            return LoadImdbGraphFromDisk(Directed);
          case Constants.IMDB_GRAPH_SOURCE_COSMOS:
            return await LoadImdbGraphFromCosmosAsync(Directed);
          default:
            _logger.LogWarning($"undefined graph source: {Source}");
            return null;
        }
      }
      catch (Exception ex)
      {
        _logger.LogError(ex, $"buildImdbGraph, exception: {ex.Message}");
        throw;
      }
    }

    private IMutableGraph<string, Edge<string>> CreateGraphObject(bool directed)
    {
      IMutableGraph<string, Edge<string>> graph;

      if (directed)
      {
        graph = new BidirectionalGraph<string, Edge<string>>();
      }
      else
      {
        graph = new UndirectedGraph<string, Edge<string>>();
      }
      return graph;
    }

    private IMutableGraph<string, Edge<string>> LoadImdbGraphFromDisk(bool directed)
    {
      throw new NotImplementedException();
      // IMutableGraph<string, Edge<string>> graph = CreateGraphObject(directed);

      // JsonLoader jsonLoader = new JsonLoader();
      // Dictionary<string, Movie> moviesHash = new Dictionary<string, Movie>();
      // jsonLoader.readMovieDocuments(moviesHash, true);
      // CheckMemory(true, true, "loadImdbGraphFromDisk - after reading movies from file");

      // Iterator<string> moviesIt = moviesHash.keySet().iterator();
      // long movieNodesCreated = 0;
      // long personNodesCreated = 0;
      // long edgesCreated = 0;

      // while (moviesIt.hasNext())
      // {
      //   string tconst = moviesIt.next();
      //   Movie movie = moviesHash.get(tconst);
      //   if (!graph.ContainsVertex(tconst))
      //   {
      //     graph.AddVertex(tconst);
      //     movieNodesCreated++;
      //   }

      //   Iterator<string> peopleIt = movie.getPeople().iterator();
      //   while (peopleIt.hasNext())
      //   {
      //     string nconst = peopleIt.next();
      //     if (!graph.ContainsVertex(nconst))
      //     {
      //       graph.AddVertex(nconst);
      //       personNodesCreated++;
      //     }
      //     graph.AddEdge(nconst, tconst);  // person-to-movie
      //     edgesCreated++;
      //     if (directed)
      //     {
      //       // just a single edge between vertices
      //     }
      //     else
      //     {
      //       graph.AddEdge(tconst, nconst);  // movie-to-person
      //       edgesCreated++;
      //     }
      //   }
      // }

      // CheckMemory(true, true, "loadImdbGraphFromDisk - after building graph");
      // _logger.LogWarning($"loadImdbGraphFromDisk - movieNodesCreated:  {movieNodesCreated}");
      // _logger.LogWarning($"loadImdbGraphFromDisk - personNodesCreated: {personNodesCreated}");
      // _logger.LogWarning($"loadImdbGraphFromDisk - edgesCreated:       {edgesCreated}");
      // return graph;
    }

    private async Task<IMutableGraph<string, Edge<string>>> LoadImdbGraphFromCosmosAsync(bool directed)
    {
      Uri = _cosmosOptions.ConnectionString;
      //Key = _cosmosOptions.Key;
      DbName = _cosmosOptions.DatabaseId;
      Source = _imdbOptions.GraphSource;
      Directed = _imdbOptions.GraphDirected;

      IMutableGraph<string, Edge<string>> graph = CreateGraphObject(directed);

      CosmosClient client;
      Database database;
      Container container;
      double requestCharge = 0;
      long documentsRead = 0;
      long movieNodesCreated = 0;
      long personNodesCreated = 0;
      long edgesCreated = 0;
      var getDataSql = new QueryDefinition("select * from c where c.pk = @pk")
        .WithParameter("@pk", Constants.DOCTYPE_MOVIE_SEED);
      int pageSize = 1000;
      string? continuationToken = null;

      _logger.LogWarning($"uri:    {Uri}");
      _logger.LogWarning($"key:    {Key}");
      _logger.LogWarning($"dbName: {DbName}");
      _logger.LogWarning($"sql:    {getDataSql.QueryText}");

      long startMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
      JsonSerializerOptions jsonSerializerOptions = new JsonSerializerOptions()
      {
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
      };
      CosmosSystemTextJsonSerializer cosmosSystemTextJsonSerializer = new CosmosSystemTextJsonSerializer(jsonSerializerOptions);
      client = new CosmosClientBuilder(
        connectionString: Uri
      )
              .WithApplicationPreferredRegions(_cosmosOptions.PreferredLocations)
              .WithConsistencyLevel(ConsistencyLevel.Session)
              .WithContentResponseOnWrite(true)
              .WithCustomSerializer(cosmosSystemTextJsonSerializer)
              .Build();

      database = client.GetDatabase(DbName);
      _logger.LogWarning($"client connected to database Id: {database.Id}");

      container = database.GetContainer(_imdbOptions.SeedContainerName);
      _logger.LogWarning($"container: {container.Id}");

      long dbConnectMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

      try
      {
        var requestOptions = new QueryRequestOptions();
        requestOptions.EnableScanInQuery = true;
        var getDataSqlCount = new QueryDefinition("select VALUE COUNT(1) from c where c.pk = @pk")
          .WithParameter("@pk", Constants.DOCTYPE_MOVIE_SEED);

        var queryIterator = container.GetItemQueryIterator<long>(getDataSqlCount);
        var response = await queryIterator.ReadNextAsync();
        long count = response.Resource.FirstOrDefault();

        _logger.LogWarning($"count: {count}");

        OnRaiseJGraphBuilderStartedLoadingProgressEvent(new JGraphBuilderStartedLoadingProgressEventArgs(count));
      }
      catch (Exception ex)
      {
        _logger.LogError(ex, "Error getting count");
      }

      try
      {
        using (FeedIterator<SeedDocument> feedResponseIterator =
                container.GetItemQueryIterator<SeedDocument>(getDataSql, continuationToken, new QueryRequestOptions { MaxItemCount = pageSize }))
        {
          do
          {
            while (feedResponseIterator.HasMoreResults)
            {
              FeedResponse<SeedDocument> page = await feedResponseIterator.ReadNextAsync();

              OnRaiseJGraphBuilderLoadingProgressEvent(new JGraphBuilderLoadingProgressEventArgs(page.Resource.Count()));

              foreach (SeedDocument doc in page.Resource)
              {
                documentsRead++;
                if ((documentsRead % 10000) == 0)
                {
                  _logger.LogWarning($"{documentsRead} -> {JsonSerializer.Serialize(doc)}");
                }
                string tconst = doc.TargetId;
                ((IMutableVertexSet<string>)graph).AddVertex(tconst);
                movieNodesCreated++;

                for (int i = 0; i < doc.AdjacentVertices.Count(); i++)
                {
                  string nconst = doc.AdjacentVertices[i];
                  if (!((IImplicitVertexSet<string>)graph).ContainsVertex(nconst))
                  {
                    ((IMutableVertexSet<string>)graph).AddVertex(nconst);
                    personNodesCreated++;
                  }
                  ((IMutableEdgeListGraph<string, Edge<string>>)graph).AddEdge(new Edge<string>(nconst, tconst));  // person-to-movie
                  edgesCreated++;

                  if (directed)
                  {
                    // just a single edge between vertices
                  }
                  else
                  {
                    ((IMutableEdgeListGraph<string, Edge<string>>)graph).AddEdge(new Edge<string>(tconst, nconst));  // movie-to-person
                    edgesCreated++;
                  }
                }
              }
              requestCharge = requestCharge + page.RequestCharge;
              continuationToken = page.ContinuationToken;

            }
          }
          while (continuationToken != null);
        }
      }
      catch (Exception ex)
      {
        //t.printStackTrace();
        _logger.LogError(ex, $"loadImdbGraphFromCosmos - exception: {ex.Message}");
      }

      long finishMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
      long dbConnectElapsed = dbConnectMs - startMs;
      long dbReadingElapsed = finishMs - dbConnectMs;
      double dbReadingSeconds = (double)dbReadingElapsed / 1000.0;
      long totalElapsed = finishMs - startMs;

      double ruPerSec = (double)requestCharge / dbReadingSeconds;

      //CheckMemory(true, true, "loadImdbGraphFromCosmos - after building graph");
      _logger.LogWarning($"loadImdbGraphFromCosmos - documentsRead:      {documentsRead}");
      _logger.LogWarning($"loadImdbGraphFromCosmos - movieNodesCreated:  {movieNodesCreated}");
      _logger.LogWarning($"loadImdbGraphFromCosmos - personNodesCreated: {personNodesCreated}");
      _logger.LogWarning($"loadImdbGraphFromCosmos - edgesCreated:       {edgesCreated}");
      _logger.LogWarning($"loadImdbGraphFromCosmos - requestCharge:      {requestCharge}");
      _logger.LogWarning($"loadImdbGraphFromCosmos - ru per second:      {ruPerSec}");
      _logger.LogWarning($"loadImdbGraphFromCosmos - db connect ms:      {dbConnectElapsed}");
      _logger.LogWarning($"loadImdbGraphFromCosmos - db read ms:         {dbReadingElapsed}");
      _logger.LogWarning($"loadImdbGraphFromCosmos - total elapsed ms:   {totalElapsed}");

      OnRaiseJGraphBuilderFinishedLoadingProgressEvent(new JGraphBuilderFinishedLoadingProgressEventArgs(documentsRead));

      return graph;
    }

    protected virtual void OnRaiseJGraphBuilderStartedLoadingProgressEvent(JGraphBuilderStartedLoadingProgressEventArgs e)
    {
      EventHandler<JGraphBuilderStartedLoadingProgressEventArgs>? raiseEvent = JGraphBuilderStartedLoadingProgress;

      if (raiseEvent != null)
      {
        raiseEvent(this, e);
      }
    }

    protected virtual void OnRaiseJGraphBuilderLoadingProgressEvent(JGraphBuilderLoadingProgressEventArgs e)
    {
      EventHandler<JGraphBuilderLoadingProgressEventArgs>? raiseEvent = JGraphBuilderLoadingProgress;

      if (raiseEvent != null)
      {
        raiseEvent(this, e);
      }
    }

    protected virtual void OnRaiseJGraphBuilderFinishedLoadingProgressEvent(JGraphBuilderFinishedLoadingProgressEventArgs e)
    {
      EventHandler<JGraphBuilderFinishedLoadingProgressEventArgs>? raiseEvent = JGraphBuilderFinishedLoadingProgress;

      if (raiseEvent != null)
      {
        raiseEvent(this, e);
      }
    }

    // protected MemoryStats CheckMemory(bool doGc, bool display, string note)
    // {
    //   if (doGc)
    //   {
    //     System.gc();
    //   }
    //   MemoryStats ms = new MemoryStats(note);
    //   if (display)
    //   {
    //     try
    //     {
    //       sysout(ms.asDelimitedHeaderLine("|"));
    //       sysout(ms.asDelimitedDataLine("|"));
    //     }
    //     catch (Exception e)
    //     {
    //       sysout("error serializing MemoryStats to JSON");
    //     }
    //   }
    //   return ms;
    // }

    // protected void sysout(string s)
    // {
    //   System.out.println(s);
    // }

  }

  public class JGraphBuilderStartedLoadingProgressEventArgs
  {
    public long MaxCount { get; private set; }
    public JGraphBuilderStartedLoadingProgressEventArgs(long maxCount)
    {
      MaxCount = maxCount;
    }
  }

  public class JGraphBuilderLoadingProgressEventArgs
  {
    public long Progress { get; private set; }
    public JGraphBuilderLoadingProgressEventArgs(long progress)
    {
      Progress = progress;
    }
  }

  public class JGraphBuilderFinishedLoadingProgressEventArgs
  {
    public long Count { get; private set; }
    public JGraphBuilderFinishedLoadingProgressEventArgs(long count)
    {
      Count = count;
    }
  }
}